ibc-testkit/tests/core/ics02_client/recover_client.rs

```diff
@@ -0,0 +1,206 @@
+use ibc::core::client::types::{msgs::MsgRecoverClient, Height};
+use ibc::core::entrypoint::{execute, validate};
+use ibc::core::handler::types::msgs::MsgEnvelope;
+use ibc::core::host::types::identifiers::ClientId;
+use ibc::core::host::ValidationContext;
+use ibc::core::primitives::Signer;
+
+use ibc_testkit::fixtures::core::signer::dummy_account_id;
+use ibc_testkit::testapp::ibc::clients::mock::client_state::{
+    client_type as mock_client_type, MockClientState,
+};
+use ibc_testkit::testapp::ibc::clients::mock::consensus_state::MockConsensusState;
+use ibc_testkit::testapp::ibc::clients::mock::header::MockHeader;
+use ibc_testkit::testapp::ibc::core::router::MockRouter;
+use ibc_testkit::testapp::ibc::core::types::MockContext;
+
+use rstest::*;
+
+struct Fixture {
+    ctx: MockContext,
+    router: MockRouter,
+    subject_client_id: ClientId,
+    substitute_client_id: ClientId,
+    signer: Signer,
+}
+
+/// Initializes the testing fixture for validating client recovery logic.
+///
+/// Creates the subject and substitute clients via sending two `MsgCreateClient`
+/// messages. The subject client is initialized in an expired state by sleeping
+/// for the duration of its trusting period. The substitute client state is
+/// initialized with a much longer trusting period and at a height higher than
+/// the subject client state's in order to ensure that it remains in an active
+/// state.
+fn setup_client_recovery_fixture(
+    subject_trusting_period: Duration,
+    subject_height: Height,
+    substitute_trusting_period: Duration,
+    substitute_height: Height,
+) -> Fixture {
+    let mut ctx = MockContext::default();
+    let mut router = MockRouter::new_with_transfer();
+    let signer = dummy_account_id();
+
+    let subject_client_state =
+        MockClientState::new(MockHeader::new(subject_height).with_current_timestamp())
+            .with_trusting_period(subject_trusting_period);
+
+    // Create the subject client
+    let msg = MsgCreateClient::new(
+        subject_client_state.into(),
+        MockConsensusState::new(MockHeader::new(subject_height).with_current_timestamp()).into(),
+        signer.clone(),
+    );
+
+    let msg_envelope = MsgEnvelope::from(ClientMsg::from(msg));
+
+    let client_type = mock_client_type();
+    let subject_client_id = client_type.build_client_id(ctx.client_counter().unwrap());
+
+    validate(&ctx, &router, msg_envelope.clone()).expect("create subject client validation");
+    execute(&mut ctx, &mut router, msg_envelope).expect("create subject client execution");
+
+    // Ensure that the subject client state is in the `Status::Expired` state
+    // by sleeping for the length of its trusting period
+    sleep(subject_trusting_period);
+
+    // Create the substitute client
+    let substitute_client_state =
+        MockClientState::new(MockHeader::new(substitute_height).with_current_timestamp())
+            .with_trusting_period(substitute_trusting_period);
```

Comment:
you may create a header with a specific timestamp by `with_timestamp`. That way we don't need to call `sleep`.

Reply:
Are you thinking of initializing the subject client with a timestamp that is equivalent to `Duration::now() - subject_trusting_period` so that we can avoid the `sleep` call?

Reply:
Yes. I think it's better to just use a single timestamp at the beginning and increment it to simulate time passing.

You should use `MockHeader::with_timestamp` in that case.

Reply:
[62f2fc7](https://github.com/cosmos/ibc-rs/pull/1151/commits/62f2fc73ab10c284673cad918a0302e20db39d84)